app/updatemanager.go

```diff
@@ -0,0 +1,163 @@
+	dbusConn                    dbus.Handle
+	updateControlTimeoutSeconds int
+}
+
+func NewUpdateManager(updateControlTimeoutSeconds int) *UpdateManager {
+	return &UpdateManager{
+		updateControlTimeoutSeconds: updateControlTimeoutSeconds,
+	}
+	return &UpdateManager{}
+}
+
+func (u *UpdateManager) EnableDBus(api dbus.DBusAPI) {
+	u.dbus = api
+}
+
+func (u *UpdateManager) Start() (context.CancelFunc, error) {
+	log.Debug("Running the UpdateManager")
+	if u.dbus == nil {
+		return nil, errors.New("DBus not enabled")
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go u.run(ctx)
+	return cancel, nil
+}
+
+func (u *UpdateManager) run(ctx context.Context) error {
+	log.Debug("Running the UpdateManager")
+	dbusConn, err := u.dbus.BusGet(dbus.GBusTypeSystem)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start the Update manager")
+	}
+	nameGID, err := u.dbus.BusOwnNameOnConnection(
+		dbusConn,
+		UpdateManagerDBusObjectName,
+		dbus.DBusNameOwnerFlagsAllowReplacement|dbus.DBusNameOwnerFlagsReplace,
+	)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start the update manager")
+	}
+	defer u.dbus.BusUnownName(nameGID)
+	intGID, err := u.dbus.BusRegisterInterface(
+		dbusConn,
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterface,
+	)
+	if err != nil {
+		log.Errorf("Failed to register the DBus interface name %q at path %q: %s",
+			UpdateManagerDBusInterface,
+			UpdateManagerDBusPath,
+			err,
+		)
+		return err
+	}
+	defer u.dbus.BusUnregisterInterface(dbusConn, intGID)
+
+	u.dbus.RegisterMethodCallCallback(
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterfaceName,
+		"SetUpdateControlMap",
+		func(objectPath string, interfaceName string, methodName string, updateControlMap string) (interface{}, error) {
+			controlMap := UpdateControlMap{}
+			err := json.Unmarshal([]byte(updateControlMap), &controlMap)
+			if err != nil {
+				log.Errorf("Failed to unmarshal the JSON in the string received via SetControlMap on D-Bus: %s", err)
+				return u.updateControlTimeoutSeconds / 2, err
+			}
+			log.Debugf("Setting the control map parameter to: %s", controlMap.ID)
+			UpdateManagerMap.Set(&controlMap)
+			return u.updateControlTimeoutSeconds / 2, nil
+		})
+	defer u.dbus.UnregisterMethodCallCallback(
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterfaceName,
+		"SetUpdateControlMap")
+	select {
+	case <-ctx.Done():
+		return nil
+	}
+	return nil
+}
+
+type UpdateControlMap struct {
+	ID       string `json:"id"`
+	Priority int    `json:"priority"`
+}
+
+func (u *UpdateControlMap) key() string {
+	return u.ID + strconv.Itoa(u.Priority)
+}
+
+type ControlMap struct {
+	controlMap map[string]*UpdateControlMap
+	mutex      sync.Mutex
+}
+
+func NewControlMap() *ControlMap {
+	return &ControlMap{
+		controlMap: make(map[string]*UpdateControlMap),
+	}
+}
+
+func (c *ControlMap) Set(cm *UpdateControlMap) {
+	log.Error("Setting..")
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.controlMap[cm.key()] = cm
```

Comment:
I see a potential problem here. Here we are storing the map with a key which is combination of the ID and priority. This is fine in itself, but we need to be able to look up the map with the ID only. The interface would be more akin to this:

```go
Set(cm *UpdateControlMap)
Get(id string) []*UpdateControlMap
```

So it would return a list of all the maps that match, with possibly several different priorities. You could make this an extra function, in addition to the original `Get`, but the client will always call the getter which takes an ID only.